### DISCUSSION TITLE
Reading Saml Security Token in Core WCF

### DISCUSSION THREAD

**Discussion by riksteiner** [CoreWCF/CoreWCF#1413] [General]
Hi,

We have the below piece of code in our service for reading saml token in WCF .Net framework

```
List<SecurityToken> securityTokenList = new List<SecurityToken>();
securityTokenList.Add(new X509SecurityToken(cert1));
securityTokenList.Add(new X509SecurityToken(cert2));

SecurityTokenResolver securityTokenResolver = SecurityTokenResolver.CreateDefaultSecurityTokenResolver(securityTokenList.AsReadOnly(), true);
return (SamlSecurityToken)WSSecurityTokenSerializer.DefaultInstance.ReadToken(reader, securityTokenResolver);
```

We wanted to convert this code to .Net Core. As part of this we just changed the namespaces to CoreWCF and got the same code compiling but while running the code we hit the below excpetion

```
System.ArgumentNullException: Value cannot be null. (Parameter 'ski')    at CoreWCF.Security.SamlSerializer.ResolveSecurityKey(SecurityKeyIdentifier ski  SecurityTokenResolver tokenResolver)    at 
CoreWCF.Security.SamlSerializer.ReadToken(XmlDictionaryReader reader  SecurityTokenSerializer keyInfoSerializer  SecurityTokenResolver outOfBandTokenResolver)    at 
CoreWCF.Security.WSSecurityJan2004.SamlTokenEntry.ReadTokenCore(XmlDictionaryReader reader  SecurityTokenResolver tokenResolver)    at 
CoreWCF.Security.WSSecurityTokenSerializer.ReadTokenCore(XmlReader reader  SecurityTokenResolver tokenResolver)    
--- End of inner exception stack trace ---    at CoreWCF.Security.WSSecurityTokenSerializer.ReadTokenCore(XmlReader reader  SecurityTokenResolver tokenResolver)    at 
CoreWCF.IdentityModel.Selectors.SecurityTokenSerializer.ReadToken(XmlReader reader  SecurityTokenResolver tokenResolver)    at
```

Any ideas on how we can resolve this issue ?

One suggestion we got was to try out is Microsoft.Identity APIs directly since CoreWCF APIs are also a thin wrapper over same for saml tokens. But we were not very sure how to proceed here since SamlSecurityTokenHandler. ReadToken(XmlReader reader, TokenValidationParameters validationParameters) is not implemented

https://github.com/ryannewington/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/0665af62cc58a28ebe184dd97f4d018f84e1d83d/src/Microsoft.IdentityModel.Tokens.Saml/Saml/SamlSecurityTokenHandler.cs#L793C9-L796C10

We can see SamlSecurityTokenHandler. ReadToken(XmlReader reader) is available but then we were not sure how to pass the cert info to the API which was done in .Net framework via SecurityTokenResolver

We have been blocked on this for a while and any pointer will be super helpful

**Comment by mconnew** [comment]
There's two different implementations wrapping the MS.IM implementation in CoreWCF. One is in the class CoreWCF.Security.SamlSerializer, and the other is in CoreWCF.IdentityModel.Tokens.SamlSecurityTokenHandler. The each do it slightly different ways.  
`CoreWCF.Security.SamlSerializer` (which I'll call SamlSerializer from here on) wraps and uses `Microsoft.IdentityModel.Tokens.Saml.SamlSerializer` (which I'll call IMSamlSerializer). SamlSerializer.ReadToken calls IMSamlSerializer.ReadAssertion passing the XmlReader which returns a Microsoft.IdentityModel.Tokens.Saml.SamlAssertion (IMSamlAssertion). It then tries to create a SecurityKeyIdentifier from IMSamlAssertion.Signature.KeyInfo. It does this by iterating over KeyInfo.X509Data (a collection of X509Data), and creating a X509RawDataKeyIdentifierClause instance from the first X509Data instance which contains any certificates. As the ski parameter being passed to ResolveSecurityKey is null, that means IMSamlAssertion.Signature.KeyInfo either contains no X509Data, or none of the X509Data contains any certificates (should be a base64 encoded binary blob).

If we take a look at SamlSecurityTokenHandler, it wraps Microsoft.IdentityModel.Tokens.Saml.SamlSecurityTokenHandler (IMSamlSecurityTokenHandler). SamlSecurityTokenHandler.ReadToken calls IMSamlSecurityTokenHandler.ReadSamlToken. I don't know which version of Microsoft.IdentityModel.Tokens.Saml you are referencing, but the version CoreWCF references has an implementation of ReadToken, it just forwards to ReadSamlToken. The difference between the two is ReadSamlToken doesn't need a cast to obtain an IMSamlSecurityToken. If you look at the implementation on the default branch (dev) [you can see it implemented here](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/dev/src/Microsoft.IdentityModel.Tokens.Saml/Saml/SamlSecurityTokenHandler.cs#L738). SamlSecurityTokenHandler then also tries to create a X509RawDataKeyIdentifierClause using the same helper class.  

There's an important difference between the two implementations. SamlSecurityTokenHandler will skip trying to resolve the signing token if the key identifier can't be found whereas SamlSerializer will throw. I took a look at this implementation compared with the one from .NET Framework and they are very different. As far as I can tell, on .NET Framework, if there are no keys found, it will throw an XmlException. On .NET Framework, there are many types of keys that can be used to sign the Saml token, but the way CoreWCF has been implemented, we only support what would be the equivalent of X509RawDataKeyIdentifierClause as we only examine KeyInfo.X509Data.Certificates which is a collection of strings. The other clause types we could potentially instantiate (on .NET Framework, not in CoreWCF), are X509IssuerSerialKeyIdentifierClause and X509SubjectKeyIdentifierClause. It's possible that the X509Data instance has an empty Certificates collection and instead is referencing the certificate by some other means such as SubjectName. In that case, I believe you would need to have machine local knowledge of the certificate based on the subject name (e.g. in your local certificate store) to be able to validate the Saml token. It's also possible a certificate isn't being used at all and instead it's using an RSA key to secure the token, which it looks like is a possibility if using JwtWebKey.  

The code implemented here was a significant rework due to not wishing to have our own implementation of SignedXml (we use the runtime crypto implementation instead which has quite a different api shape) or have our own implementation of SamlSecurityTokenHandler/SamlSerializer. When this area was reimplemented, the focus was on enabling targeted scenarios. We tested against a couple of publicly available STS implementations and made sure we had what was needed for those, with any gaps filled in on a demand basis. One of two things is happening here. Either you are using an old version of the M.IM Saml package which isn't deserializing things correctly (possible as you linked to a really old git hash with code missing), or the type/nature of the key info being sent has changed and CoreWCF isn't properly supporting it. Can you provide the (data redacted) KeyInfo part of a failing Saml security token. I should be able to work out what's going on from there.

**Comment by svengeance** [comment]
@mconnew I wasn't sure if I should open a new discussion since mine is very relevant - I'm looking at being able to deserialize a KeyInfo with both RSA and x509 parts **server-side.** I've successfully implemented this behavior **client-side** (with `dotnet/wcf` modern package) by overriding `MSIM.Saml2SecurityTokenHandler CreateSubject(SecurityTokenDescriptor)` and ensuring my Saml2SubjectConfirmationData has a KeyInfo with both RSA/x509 parts.

I anticipate on the server side I'll need to override some minor parts of `CoreWCF.IM.Tokens.Saml2SecurityTokenHandler` - especially `TryResolveIssuerToken`, as that is currently throwing the minute it encounters an RSA block in the KeyInfo (more specifically - `CoreWCF.Security.SecurityUtils.CreateSecurityKeyIdentifier(assertion.Signature.KeyInfo)`).

Does this sound about right? Is there any ongoing effort to natively support RSA that I can look forward to? I didn't see any issues or such regarding it, so I'm worried I'm off-base.

**Reply by mconnew** [reply]
There are multiple things you could be looking to do and the answer will depend on what you're bigger picture goal is.
Do you need to verify the signature or are you just deserializing?
Do you need to validate the RSA token with something like a SevurityTokenAuthenticator? 
Do you need a WCF/CoreWCF OK class like RSASecurityToken to wrap the RSA key material or are you just wanting the raw RSA key? 
Is this part of authentication for a CoreWCF service?

**Reply by svengeance** [reply]
Your thoroughness and vigor for this project is commendable, man. 
Forgive me, I'm picking up the source material as I port it, and I hope it doesn't sound too heretical.

We've overridden the AuthorizationManager to always return true, relying solely on SAML (token) and Certificate (transport-secured) validation. Certs are required for incoming requests, and have to match trusted ones on the machine/store.
That aside, SecurityTokenAuthenticator is also overridden. It provides a custom Serializer/Saml2SecurityTokenHandler (and `ResolveSecurityKeys` impl), and a custom SecurityTokenAuthenticator which is responsible for storing the claims in the OperationContext to be ripped out at a later time.

To answer your questions directly:
1. Need to verify the x509 data matches a trusted certificate
2. The RSA portion of the token isn't validated by _our_ code, but maybe it is validated by one of .NET Framework's WCF pieces? This is where I'm a little fuzzy. I know FW and System.IdentityModel has more support for various keys, so I can only assume that I will need to validate it myself in CoreWCF. I believe this is done by validating all 3 bits of data in the KeyInfo (Cert, RsaModulus, RsaExponent) against the local trusted cert?
3. I don't need the key after validation or need any wrapper for it, just to be able to consume and validate it
4. Yes, part of authn. The authz portion is entirely done via SamlAssertions/claims.

I know a ~picture~ message is worth a thousand words, so I've attached it if you're curious what I'm trying to port. This was produced by the current (NetFramework) system, and is something I **can** reproduce as a client with System.ServiceModel on .NET Core. The final step is parsing it correctly with CoreWCF.

<details>
<summary>Xml</summary>

```xml
<Envelope xmlns:a="http://www.w3.org/2005/08/addressing"
	xmlns:u="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-utility-1.0.xsd"
	xmlns:s="http://www.w3.org/2003/05/soap-envelope">
	<Header>
		<Security s:mustUnderstand="1"
			xmlns:o="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-secext-1.0.xsd">
			<Timestamp u:Id="_0">
				<Created>
					2025-09-03T19:45:01.246Z
				</Created>
				<Expires>
					2025-09-03T19:50:01.246Z
				</Expires>
			</Timestamp>
			<Assertion ID="_14d242ba-7a90-4ada-9c02-640087b62b3b" IssueInstant="2025-09-03T19:45:01.105Z" Version="2.0"
				xmlns="urn:oasis:names:tc:SAML:2.0:assertion">
				<Issuer Format="urn:oasis:names:tc:SAML:1.1:nameid-format:X509SubjectName"> CN=foo.com, OU=foo, O=FOO LLC, L=Foo, S=Foo, C=Foo
				</Issuer>
				<Signature xmlns="http://www.w3.org/2000/09/xmldsig#">
					<SignedInfo>
						<CanonicalizationMethod Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#" />
						<SignatureMethod Algorithm="http://www.w3.org/2001/04/xmldsig-more#rsa-sha256" />
						<Reference URI="#_14d242ba-7a90-4ada-9c02-640087b62b3b">
							<Transforms>
								<Transform Algorithm="http://www.w3.org/2000/09/xmldsig#enveloped-signature" />
								<Transform Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#" />
							</Transforms>
							<DigestMethod Algorithm="http://www.w3.org/2001/04/xmlenc#sha256" />
							<DigestValue> ...
							</DigestValue>
						</Reference>
					</SignedInfo>
					<SignatureValue> 1p...
					</SignatureValue>
					<KeyInfo>
						<X509Data>
							<X509Certificate> MII...
							</X509Certificate>
						</X509Data>
						<KeyValue>
							<RSAKeyValue>
								<Modulus> 6t...
								</Modulus>
								<Exponent>
									...
								</Exponent>
							</RSAKeyValue>
						</KeyValue>
					</KeyInfo>
				</Signature>
				<Subject>
					<NameID Format="urn:oasis:names:tc:SAML:1.1:nameid-format:emailAddress">
						...
					</NameID>
					<SubjectConfirmation Method="urn:oasis:names:tc:SAML:2.0:cm:holder-of-key">
						<SubjectConfirmationData a:type="KeyInfoConfirmationDataType"
							xmlns:a="http://www.w3.org/2001/XMLSchema-instance">
							<KeyInfo xmlns="http://www.w3.org/2000/09/xmldsig#">
								<X509Data>
									<X509Certificate> MII...
									</X509Certificate>
								</X509Data>
								<KeyValue>
									<RSAKeyValue>
										<Modulus> ...
										</Modulus>
										<Exponent>
											...
										</Exponent>
									</RSAKeyValue>
								</KeyValue>
							</KeyInfo>
						</SubjectConfirmationData>
					</SubjectConfirmation>
				</Subject>
				<Conditions NotBefore="2025-09-03T19:45:01.105Z" NotOnOrAfter="2025-09-04T03:45:01.105Z">
					<AudienceRestriction>
						<Audience>
							https://...
						</Audience>
					</AudienceRestriction>
				</Conditions>
				<AuthnStatement AuthnInstant="2025-09-03T19:45:01.105Z">
					<AuthnContext>
						<AuthnContextClassRef>
							urn:oasis:names:tc:SAML:2.0:ac:classes:PasswordProtectedTransport
						</AuthnContextClassRef>
					</AuthnContext>
				</AuthnStatement>
				<AttributeStatement>
				</AttributeStatement>
			</Assertion>
			<Signature xmlns="http://www.w3.org/2000/09/xmldsig#">
				<SignedInfo>
					<CanonicalizationMethod Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#" />
					<SignatureMethod Algorithm="http://www.w3.org/2001/04/xmldsig-more#rsa-sha256" />
					<Reference URI="#_0">
						<Transforms>
							<Transform Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#" />
						</Transforms>
						<DigestMethod Algorithm="http://www.w3.org/2001/04/xmlenc#sha256" />
						<DigestValue> Vc...
						</DigestValue>
					</Reference>
					<Reference URI="#_1">
						<Transforms>
							<Transform Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#" />
						</Transforms>
						<DigestMethod Algorithm="http://www.w3.org/2001/04/xmlenc#sha256" />
						<DigestValue> kq...
						</DigestValue>
					</Reference>
				</SignedInfo>
				<SignatureValue> 1dXa...
				</SignatureValue>
					<KeyInfo>
						<SecurityTokenReference p9:TokenType="http://docs.oasis-open.org/wss/oasis-wss-saml-token-profile-1.1#SAMLV2.0"
							xmlns:p9="http://docs.oasis-open.org/wss/oasis-wss-wssecurity-secext-1.1.xsd">
							<KeyIdentifier ValueType="http://docs.oasis-open.org/wss/oasis-wss-saml-token-profile-1.1#SAMLID">
								_14d242ba-7a90-4ada-9c02-640087b62b3b
							</KeyIdentifier>
						</SecurityTokenReference>
					</KeyInfo>
			</Signature>
		</Security>
	</Header>
</Envelope>
```

</details>

**Reply by svengeance** [reply]
I've got what I believe is a basic.. implementation of this - I'm now getting hung up on `ReceiveSecurityHeader.ProcessSupportingSignatureAsync` and `GetSupportingTokenTracker()` inside. My `rootSigningToken` is an x509SecurityToken, but the token available in `_supportingTokenTrackers` is a `Saml2SecurityToken` which **contains** an x509Securitytoken inside as the SigningToken.

Am I on a wrong path which resulted in the signing token being an x509SecurityToken?

<img width="2551" height="1288" alt="image" src="https://github.com/user-attachments/assets/cd129499-1301-41e2-9e90-b77f1cc2a08d" />
